feat: add demand flexibility cost

dongqi-wu opened the pull request:
[Pull Request doc](https://breakthrough-energy.github.io/docs/user/git_guide.html#d-pull-request)

### Purpose
Update demand flexibility cost code in #135 to accommodate the changes in #142.  Consolidate comments from @lanesmith 
Respond to comments in PR #147 (closed accidentally)

### What the code is doing
* The code attempts to read two additional input files `demand_flexibility_cost_up.csv` and `demand_flexibility_cost_up.csv` and create corresponding fields in the DemandFlexibility class
* The numeric costs are added to the decision variables `load_shift_up` and `load_shift_dn`

### Testing
The code is setup on a local computer and uses a ERCOT sub-system to reduce computation time
The code is tested using dummy input files (shared in slack channel) and when inputs files are missing
The results with/without the cost input files are checked and compared
Various testing are performed using interval length and interval number ranging from 1 to 3 

### Where to look
* It's helpful to clarify where your new code lives if you moved files around or there could be confusion/

* two fields in types.jl
* csv reading code in read.jl
* add corresponding objective function expression in model.jl
* re-assign coefficients in loop.jl

### Usage Example/Visuals
N/A

### Time estimate
~15 min for people that are familiar with the previous version in #147

danielolsen commented on src/model.jl:
```diff
@@ -492,12 +492,15 @@ function _add_objective_function!(
     storage_e0::Array{Float64,1},
-)
+    demand_flexibility::DemandFlexibility,
+    )
```
Please de-indent this parenthesis.

dongqi-wu commented:
@danielolsen @lanesmith 
Sorry I somehow accidentally closed the last PR (it seems git was asking me to pull the head before I push my interactive rebase, then I might have pushed a version that is exactly the same as the head which caused the last PR to close..) 
I have addressed the comments from #147, please check this version.

danielolsen commented on src/read.jl:
```diff
@@ -192,6 +194,21 @@ function read_demand_flexibility(filepath, interval)::DemandFlexibility
         end
+        
+        # Try reading the cost for up- and down-shifting loads
+        for cost_file_postfix in ("up", "dn")
+            try 
+                demand_flexibility["cost_$cost_file_postfix"] = CSV.File(
+                    joinpath(filepath, "demand_flexibility_cost_$cost_file_postfix.csv")
+                ) |> DataFrames.DataFrame
+                println("...loading demand flexibility $cost_file_postfix-shift cost profiles")
+            catch e
+                println(
+                    "Demand flexibility down-shift cost profiles not found in " * filepath
+                )
+                demand_flexibility["cost_$cost_file_postfix"] = nothing
```
What does `cost_file_postfix` mean?

dongqi-wu replied:
> What does `cost_file_postfix` mean?

Oh I meant "suffix".. the conterpart is "prefix". I need to review my english,

danielolsen replied:
Aha! Don't worry, English does not make any sense.

dongqi-wu replied:
> Aha! Don't worry, English does not make any sense.

I have fixed locally. I will wait for the other comments and update later.

danielolsen replied:
The code looks good to me, and @lanesmith suggested in the old PR that once his comments are addressed (which it seem like they are now) that it's ready to go. If you push the small changes requested on this one, I will approve and we can get this one merged, and then we will rebase #145 and merge that one as well.

dongqi-wu replied:
> The code looks good to me, and @lanesmith suggested in the old PR that once his comments are addressed (which it seem like they are now) that it's ready to go. If you push the small changes requested on this one, I will approve and we can get this one merged, and then we will rebase #145 and merge that one as well.

I just pushed through interactive rebase.. why there are the strange log texts in the source code?

BainanXia replied:
@dongqi-wu I think you need to delete those texts manually when resolving conflicts. Those are the markups that Github automatically generated when conflicts are detected, which tells you what are the conflicts explicitly.

danielolsen replied:
It looks like there were some merge conflicts that got committed before being resolved. Would you like me to force-push a fixed commit with a better commit message into your repo?

dongqi-wu replied:
> It looks like there were some merge conflicts that got committed before being resolved. Would you like me to force-push a fixed commit with a better commit message into your repo?

Hi Daniel I just saw the message.. I was trying to fix myself and it seems there are some problems with the commits (as of now). The changes seem to be separated into the 2 commits and when I do interactive rebase and drop the 1st one it only keep the change in the 2nd one. I think I am supposed to make 1 commit that is the union the current 2 commits. Is there something I did wrong?

danielolsen replied:
You didn't want to `drop`, you wanted to designate the second commit as a `fixup`, which will combine that commit's changes with the commit before. Next time you have a git issue, send us a message on Slack, that should be the fastest way to resolve it.

dongqi-wu committed:
feat: add demand flexibility cost

BainanXia commented:
I've read through the code which makes sense to me. Just curious how do we resolve the issue @lanesmith mentioned during last meeting: `set_objective_coefficient` works on a toy JuMP model but throws following error when calling it in `loop.jl`?

```julia
JULIA: MathOptInterface.ModifyObjectiveNotAllowed{MathOptInterface.ScalarCoefficientChange{Int64}}: Modifying the objective function with MathOptInterface.ScalarCoefficientChange{Int64}(MathOptInterface.VariableIndex(195001), 5) cannot be performed. You may want to use a `CachingOptimizer` in `AUTOMATIC` mode or you may need to call `reset_optimizer` before doing this operation if the `CachingOptimizer` is in `MANUAL` mode.
```

dongqi-wu committed:
feat: add demand flexibility cost

lanesmith commented:
> I've read through the code which makes sense to me. Just curious how do we resolve the issue @lanesmith mentioned during last meeting: `set_objective_coefficient` works on a toy JuMP model but throws following error when calling it in `loop.jl`?
> 
> ```julia
> JULIA: MathOptInterface.ModifyObjectiveNotAllowed{MathOptInterface.ScalarCoefficientChange{Int64}}: Modifying the objective function with MathOptInterface.ScalarCoefficientChange{Int64}(MathOptInterface.VariableIndex(195001), 5) cannot be performed. You may want to use a `CachingOptimizer` in `AUTOMATIC` mode or you may need to call `reset_optimizer` before doing this operation if the `CachingOptimizer` is in `MANUAL` mode.
> ```

When I tested on the new laptop with a new environment, the implementation that we had for `loop.jl` seemed to work. I'm not sure what might have caused the problem for me in the past, but perhaps it was an older `JuMP` version? I'm not confident that was the problem though as the `Project.toml` file should have been up-to-date on my end

danielolsen commented:
> When I tested on the new laptop with a new environment, the implementation that we had for loop.jl seemed to work. I'm not sure what might have caused the problem for me in the past, but perhaps it was an older JuMP version? I'm not confident that was the problem though as the Project.toml file should have been up-to-date on my end

By default, bug-fix versions are not pinned, so if we specified an older version of JuMP, that should not prevent a newer, compatible bug-fix version from being installed instead.

BainanXia commented:
> > I've read through the code which makes sense to me. Just curious how do we resolve the issue @lanesmith mentioned during last meeting: `set_objective_coefficient` works on a toy JuMP model but throws following error when calling it in `loop.jl`?
> > ```julia
> > JULIA: MathOptInterface.ModifyObjectiveNotAllowed{MathOptInterface.ScalarCoefficientChange{Int64}}: Modifying the objective function with MathOptInterface.ScalarCoefficientChange{Int64}(MathOptInterface.VariableIndex(195001), 5) cannot be performed. You may want to use a `CachingOptimizer` in `AUTOMATIC` mode or you may need to call `reset_optimizer` before doing this operation if the `CachingOptimizer` is in `MANUAL` mode.
> > ```
> 
> When I tested on the new laptop with a new environment, the implementation that we had for `loop.jl` seemed to work. I'm not sure what might have caused the problem for me in the past, but perhaps it was an older `JuMP` version? I'm not confident that was the problem though as the `Project.toml` file should have been up-to-date on my end

Aha, glad to hear it is resolved magically.
